File: examples/python/token-ft/FungibleTokenExample.runar.py
from runar import (
    StatefulSmartContract, PubKey, Sig, ByteString, Bigint, Readonly,
    public, assert_, check_sig,
)

class FungibleToken(StatefulSmartContract):
    owner: PubKey
    balance: Bigint
    token_id: Readonly[ByteString]

    def __init__(self, owner: PubKey, balance: Bigint, token_id: ByteString):
        super().__init__(owner, balance, token_id)
        self.owner = owner
        self.balance = balance
        self.token_id = token_id

    @public
    def transfer(self, sig: Sig, to: PubKey, amount: Bigint, output_satoshis: Bigint):
        assert_(check_sig(sig, self.owner))
        assert_(amount > 0)
        assert_(amount <= self.balance)
        self.add_output(output_satoshis, to, amount)
        self.add_output(output_satoshis, self.owner, self.balance - amount)

    @public
    def send(self, sig: Sig, to: PubKey, output_satoshis: Bigint):
        assert_(check_sig(sig, self.owner))
        self.add_output(output_satoshis, to, self.balance)

    @public
    def merge(self, sig: Sig, total_balance: Bigint, output_satoshis: Bigint):
        assert_(check_sig(sig, self.owner))
        assert_(total_balance >= self.balance)
        self.add_output(output_satoshis, self.owner, total_balance)

File: examples/ruby/token-ft/FungibleTokenExample.runar.rb
require 'runar'

class FungibleToken < Runar::StatefulSmartContract
  prop :owner, PubKey
  prop :balance, Bigint
  prop :token_id, ByteString, readonly: true

  def initialize(owner, balance, token_id)
    super(owner, balance, token_id)
    @owner = owner
    @balance = balance
    @token_id = token_id
  end

  runar_public sig: Sig, to: PubKey, amount: Bigint, output_satoshis: Bigint
  def transfer(sig, to, amount, output_satoshis)
    assert check_sig(sig, @owner)
    assert amount > 0
    assert amount <= @balance
    add_output(output_satoshis, to, amount)
    add_output(output_satoshis, @owner, @balance - amount)
  end

  runar_public sig: Sig, to: PubKey, output_satoshis: Bigint
  def send(sig, to, output_satoshis)
    assert check_sig(sig, @owner)
    add_output(output_satoshis, to, @balance)
  end

  runar_public sig: Sig, total_balance: Bigint, output_satoshis: Bigint
  def merge(sig, total_balance, output_satoshis)
    assert check_sig(sig, @owner)
    assert total_balance >= @balance
    add_output(output_satoshis, @owner, total_balance)
  end
end
